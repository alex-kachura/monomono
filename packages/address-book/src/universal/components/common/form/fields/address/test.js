import React from 'react';
import { render, fireEvent, cleanup, waitForDomChange } from 'react-testing-library';
import { renderProviders } from '@oneaccount/react-foundations/lib/test-utils';
import { DefaultThemeProvider } from '@beans/theme';
import Address from './';
import { Formik } from '@oneaccount/react-foundations';

const noop = jest.fn();

const fields = [
  {
    id: 'postcode',
    name: 'postcode',
    label: 'Postcode',
    type: 'text',
    xtype: 'address',
    valuePath: 'postcode',
    placeholder: 'e.g. EC1R 2NT',
    required: true,
  },
];

function renderAddress({ values, fetchFn }) {
  return render(
    renderProviders({
      appConfig: {
        getLocalePhrase: (key) => key,
        fetch: fetchFn,
        region: 'en-GB',
        config: {
          'en-GB': {
            externalApps: {
              postcodeLookup: 'https://www-local.tesco.com/account/en-GB/address/addresses',
            },
          },
        },
      },
      children: (
        <DefaultThemeProvider>
          <Formik initialValues={values}>
            <Address fields={fields} />
          </Formik>
        </DefaultThemeProvider>
      ),
    }),
  );
}

describe('[Component: Address]', () => {
  afterEach(cleanup);
  describe('[State: Default]', () => {
    it('should render correctly', () => {
      const component = renderAddress({ values: { postcode: '' } });

      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe('[State: Initial Values]', () => {
    it('should render correctly', () => {
      const component = renderAddress({
        values: {
          postcode: 'EC1R 2NT',
        },
      });

      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe('[State: Initial Values]', () => {
    it('should render correctly', () => {
      const component = renderAddress({
        values: {
          postcode: '',
        },
      });

      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe('[Find Address]', () => {
    it('should call fetch and render address list', async () => {
      const fetch = jest.fn(async () => [
        {
          id: 'trn:tesco:address:address:uuid:7634e35e-70d1-4a2e-a3a3-020a3bb8b37e',
          addressLines: [
            { lineNumber: 1, value: 'Barker & Stonehouse Ltd' },
            { lineNumber: 2, value: 'Strawberry Buildings' },
            { lineNumber: 3, value: 'Leazes Park Road' },
          ],
          postcode: 'NE1 4PQ',
          postTown: 'NEWCASTLE UPON TYNE',
        },
      ]);
      const component = renderAddress({
        values: {
          postcode: '',
        },
        fetchFn: fetch,
      });

      const input = component.container.querySelector('input[name=postcode]');
      const button = component.container.querySelector('input[name=postcode] + button');

      fireEvent.change(input, {
        persist: noop,
        target: {
          name: 'postcode',
          value: 'NE1 4PQ',
        },
      });

      fireEvent.click(button, {});

      expect(fetch).toBeCalledWith(expect.stringContaining('?postcode=NE14PQ'), undefined);

      await waitForDomChange();

      expect(component.asFragment()).toMatchSnapshot();
    });
  });
});
