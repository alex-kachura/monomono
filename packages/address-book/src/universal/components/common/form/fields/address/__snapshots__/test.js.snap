// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component: Address] [Find Address] should call fetch and render address list 1`] = `
<DocumentFragment>
  <div
    class="address is-find-address"
  >
    <div
      class="address-field postcode sc-ifAKCX eKJWeb styled__Container-sc-1bf1p22-0 hTDNaU"
    >
      <label
        class="styled__StyledLabel-sc-28yvb4-0 bPkari"
        for="postcode"
      >
        Postcode 
      </label>
      <div
        class="beans-form-group__children styled__FormGroupChildren-wt4ift-2 bNvdOT"
      >
        <input
          aria-disabled="false"
          class="styled__Container-sc-1bf1p22-0 eZwXEO"
          id="postcode"
          name="postcode"
          placeholder="e.g. EC1R 2NT"
          type="text"
          value="NE1 4PQ"
        />
        <button
          aria-disabled="false"
          class="find-address styled__PrimaryButton-sc-3mnirm-2 giYBMC"
          role="button"
          type="button"
        >
          <span
            class="styled__InnerContainer-sc-3mnirm-1 fUwmia"
          >
            <span
              class="beans-button__text styled__TextSpan-sc-3mnirm-5 eqKhmg"
            >
              address.find-address
            </span>
          </span>
        </button>
      </div>
    </div>
    <div
      class="address-dropdown styled__Container-sc-1bf1p22-0 fGXsaT"
    >
      <select
        aria-disabled="false"
        class="beans-dropdown__select styled__HiddenSelect-sc-1amjgdg-2 kZoLLQ"
        role="listbox"
      >
        <option
          value="-1"
        >
          address.select-address
        </option>
        <option
          value="trn:tesco:address:address:uuid:7634e35e-70d1-4a2e-a3a3-020a3bb8b37e"
        >
          Barker & Stonehouse Ltd, Strawberry Buildings, Leazes Park Road
        </option>
      </select>
      <span
        class="beans-dropdown__select-span styled__SelectSpan-sc-1amjgdg-1 iWDBSD"
      >
        address.select-address
      </span>
      <svg
        aria-hidden="true"
        class="beans-dropdown__select-icon styled__AlignedIcon-sc-1amjgdg-0 hkwswh styled__Container-sc-1bf1p22-0 gJOjrx"
        color="currentColor"
        height="12"
        preserveAspectRatio="xMinYMax meet"
        viewBox="0 -1 20 12"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18.652 1.312L10 10.276 1.356 1.312"
          fill="none"
          stroke="#666666"
        />
      </svg>
    </div>
    <button
      aria-disabled="false"
      class="manual-link styled__LinkButton-sc-3mnirm-13 gyNqJj"
      role="button"
      type="button"
    >
      <span
        class="styled__InnerContainer-sc-3mnirm-1 fUwmia"
      >
        <span
          class="beans-button__text styled__TextSpan-sc-3mnirm-5 eqKhmg"
        >
          address.address-manually
        </span>
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`[Component: Address] [State: Default] should render correctly 1`] = `
<DocumentFragment>
  <div
    class="address is-find-address"
  >
    <div
      class="address-field postcode sc-ifAKCX eKJWeb styled__Container-sc-1bf1p22-0 hTDNaU"
    >
      <label
        class="styled__StyledLabel-sc-28yvb4-0 bPkari"
        for="postcode"
      >
        Postcode 
      </label>
      <div
        class="beans-form-group__children styled__FormGroupChildren-wt4ift-2 bNvdOT"
      >
        <input
          aria-disabled="false"
          class="styled__Container-sc-1bf1p22-0 eZwXEO"
          id="postcode"
          name="postcode"
          placeholder="e.g. EC1R 2NT"
          type="text"
          value=""
        />
        <button
          aria-disabled="false"
          class="find-address styled__PrimaryButton-sc-3mnirm-2 giYBMC"
          role="button"
          type="button"
        >
          <span
            class="styled__InnerContainer-sc-3mnirm-1 fUwmia"
          >
            <span
              class="beans-button__text styled__TextSpan-sc-3mnirm-5 eqKhmg"
            >
              address.find-address
            </span>
          </span>
        </button>
      </div>
    </div>
    <button
      aria-disabled="false"
      class="manual-link styled__LinkButton-sc-3mnirm-13 gyNqJj"
      role="button"
      type="button"
    >
      <span
        class="styled__InnerContainer-sc-3mnirm-1 fUwmia"
      >
        <span
          class="beans-button__text styled__TextSpan-sc-3mnirm-5 eqKhmg"
        >
          address.address-manually
        </span>
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`[Component: Address] [State: Initial Values] should render correctly 1`] = `
<DocumentFragment>
  <div
    class="address"
  >
    <div
      class="address-field postcode sc-ifAKCX eKJWeb styled__Container-sc-1bf1p22-0 hTDNaU"
    >
      <label
        class="styled__StyledLabel-sc-28yvb4-0 bPkari"
        for="postcode"
      >
        Postcode 
      </label>
      <div
        class="beans-form-group__children styled__FormGroupChildren-wt4ift-2 bNvdOT"
      >
        <input
          aria-disabled="false"
          class="styled__Container-sc-1bf1p22-0 eZwXEO"
          id="postcode"
          name="postcode"
          placeholder="e.g. EC1R 2NT"
          type="text"
          value="EC1R 2NT"
        />
        <button
          aria-disabled="false"
          class="find-address styled__PrimaryButton-sc-3mnirm-2 giYBMC"
          role="button"
          type="button"
        >
          <span
            class="styled__InnerContainer-sc-3mnirm-1 fUwmia"
          >
            <span
              class="beans-button__text styled__TextSpan-sc-3mnirm-5 eqKhmg"
            >
              address.find-address
            </span>
          </span>
        </button>
      </div>
    </div>
    <button
      aria-disabled="false"
      class="manual-link styled__LinkButton-sc-3mnirm-13 gyNqJj"
      role="button"
      type="button"
    >
      <span
        class="styled__InnerContainer-sc-3mnirm-1 fUwmia"
      >
        <span
          class="beans-button__text styled__TextSpan-sc-3mnirm-5 eqKhmg"
        >
          address.address-manually
        </span>
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`[Component: Address] [State: Initial Values] should render correctly 2`] = `
<DocumentFragment>
  <div
    class="address is-find-address"
  >
    <div
      class="address-field postcode sc-ifAKCX eKJWeb styled__Container-sc-1bf1p22-0 hTDNaU"
    >
      <label
        class="styled__StyledLabel-sc-28yvb4-0 bPkari"
        for="postcode"
      >
        Postcode 
      </label>
      <div
        class="beans-form-group__children styled__FormGroupChildren-wt4ift-2 bNvdOT"
      >
        <input
          aria-disabled="false"
          class="styled__Container-sc-1bf1p22-0 eZwXEO"
          id="postcode"
          name="postcode"
          placeholder="e.g. EC1R 2NT"
          type="text"
          value=""
        />
        <button
          aria-disabled="false"
          class="find-address styled__PrimaryButton-sc-3mnirm-2 giYBMC"
          role="button"
          type="button"
        >
          <span
            class="styled__InnerContainer-sc-3mnirm-1 fUwmia"
          >
            <span
              class="beans-button__text styled__TextSpan-sc-3mnirm-5 eqKhmg"
            >
              address.find-address
            </span>
          </span>
        </button>
      </div>
    </div>
    <button
      aria-disabled="false"
      class="manual-link styled__LinkButton-sc-3mnirm-13 gyNqJj"
      role="button"
      type="button"
    >
      <span
        class="styled__InnerContainer-sc-3mnirm-1 fUwmia"
      >
        <span
          class="beans-button__text styled__TextSpan-sc-3mnirm-5 eqKhmg"
        >
          address.address-manually
        </span>
      </span>
    </button>
  </div>
</DocumentFragment>
`;
